sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/ColumnDefinition.scala

```diff
@@ -162,43 +163,27 @@ object ColumnDefinition {
           col.defaultValue.foreach { default =>
             checkDefaultColumnConflicts(col)
             validateDefaultValueExpr(default, statement, col.name, col.dataType)
-            if (!default.deterministic) {
-              throw QueryCompilationErrors.defaultValueNonDeterministicError(
-                "CREATE TABLE", col.name, default.originalSQL)
-            }
           }
         }
 
       case cmd: AddColumns if cmd.columnsToAdd.exists(_.default.isDefined) =>
-        // Wrap analysis errors for default values in a more user-friendly message.
+        val statement = "ALTER TABLE ADD COLUMNS"
         cmd.columnsToAdd.foreach { c =>
           c.default.foreach { d =>
-            if (!d.resolved) {
-              throw QueryCompilationErrors.defaultValuesUnresolvedExprError(
-                "ALTER TABLE", c.colName, d.originalSQL, null)
-            }
-            validateDefaultValueExpr(d, "ALTER TABLE", c.colName, c.dataType)
-            if (!d.deterministic) {
-              throw QueryCompilationErrors.defaultValueNonDeterministicError(
-                "ALTER TABLE", c.colName, d.originalSQL)
-            }
+            validateDefaultValueExpr(d, statement, c.colName, c.dataType)
           }
         }
 
       case cmd: AlterColumns if cmd.specs.exists(_.newDefaultExpression.isDefined) =>
-        // Wrap analysis errors for default values in a more user-friendly message.
+        val statement = "ALTER TABLE ALTER COLUMN"
         cmd.specs.foreach { c =>
           c.newDefaultExpression.foreach { d =>
+            // Eagerly check resolved, as accessing datatype requires it to be resolved
```

Comment:
the alter column object is a bit different than the rest.  If i dont eagerly check here check if d is resolved and let it into the method, the d.dataType itself throws another UnresolvedError, in the cases where the default value is an unresvoled expression like 'badvalue'

Reply:
how about we make the `targetType` parameter of `def validateDefaultValueExpr` to be `Option[DataType]`? For `AlterColumns` command, we don't really need to check type castability, as we use the expression's own data type as the target type.

Reply:
makes sense, done